ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala

```diff
@@ -24,111 +21,166 @@ object MetricHandle {
+
+    def prefix: MetricName
 
-    def gauge[T <: codahale.Gauge[M], M](
+    def registry: codahale.MetricRegistry
+
+    def timer(name: MetricName): Timer = DropwizardTimer(name, registry.timer(name))
+
+    def gauge[T](name: MetricName, initial: T): Gauge[T] =
+      synchronized {
+        registry.remove(name)
+        val registeredGauge = registry.register(name, Gauges.VarGauge(initial))
+        DropwizardGauge(name, registeredGauge)
+      }
+
+    def gaugeWithSupplier[T](
         name: MetricName,
-        registerGauge: => T,
-        resetGauge: (T => Unit),
-    ): Gauge[T, M] = blocking {
+        gaugeSupplier: () => () => T,
+    ): Unit =
       synchronized {
         registry.remove(name)
-        val res: Gauge[T, M] = {
-          val gauge = registerGauge
-          Gauge(name, gauge)
-        }
-        resetGauge(res.metric)
-        res
+        val _ = registry.gauge(
+          name,
+          () => {
+            val valueGetter = gaugeSupplier()
+            new codahale.Gauge[T] { override def getValue: T = valueGetter() }
+          },
+        )
+        ()
       }
-    }
 
     def meter(name: MetricName): Meter = {
       // This is idempotent
-      Meter(name, registry.meter(name))
+      DropwizardMeter(name, registry.meter(name))
     }
 
     def counter(name: MetricName): Counter = {
       // This is idempotent
-      Counter(name, registry.counter(name))
+      DropwizardCounter(name, registry.counter(name))
     }
 
     def histogram(name: MetricName): Histogram = {
-      Histogram(name, registry.histogram(name))
+      DropwizardHistogram(name, registry.histogram(name))
     }
 
   }
 
-  trait FactoryWithDBMetrics extends MetricHandle.Factory {
+  trait FactoryWithDBMetrics extends MetricHandle.DropwizardFactory {
     def createDbMetrics(name: String): DatabaseMetrics =
       new DatabaseMetrics(prefix, name, registry)
   }
 
-  sealed case class Timer(name: String, metric: codahale.Timer)
-      extends MetricHandle[codahale.Timer] {
+  sealed trait Timer extends MetricHandle {
+
     def metricType: String = "Timer"
 
-    def timeEitherT[E, A](ev: EitherT[Future, E, A]): EitherT[Future, E, A] = {
-      EitherT(Timed.future(metric, ev.value))
+    def update(duration: Long, unit: TimeUnit): Unit
+
+    def time[T](call: => T): T
+
+    def startAsync(): TimerStop
+
+    def timeFuture[T](call: => Future[T]): Future[T] = {
+      val stop = startAsync()
+      val result = call
+      result.onComplete(_ => stop())(ExecutionContext.parasitic)
+      result
     }
+  }
 
-    def update(duration: Long, unit: TimeUnit): Unit = metric.update(duration, unit)
-    def getCount: Long = metric.getCount
-    def getSnapshot: Snapshot = metric.getSnapshot
-    def getMeanRate: Double = metric.getMeanRate
-    def time(): Context = metric.time()
+  object Timer {
+    type TimerStop = () => Unit
   }
 
-  sealed case class Gauge[U <: codahale.Gauge[T], T](name: String, metric: U)
```

Comment:
This change also breaks canton e.g. https://github.com/DACH-NY/canton/blob/e888db5dc5ab7dccba61ebef171fb2e5c46b4d3d/community/common/src/main/scala/com/digitalasset/canton/crypto/store/db/DbCryptoPrivateStore.scala#L70.

Reply:
We're mixing concepts in a pretty difficult way here. The `TimedLoadGoage` should not exist, instead this should be an extension for `Timer` that adds cats ways to work with cats.